sdk/azcore/arm/resource_identifier.go

```diff
@@ -0,0 +1,315 @@
+		if lowerFirstPart == subscriptionsKey || lowerFirstPart == resourceGroupsLowerKey {
+			return nil, fmt.Errorf("invalid resource id: %s", id)
+		}
+
+		// resourceGroup must contain either child or provider resource type
+		if parent.resourceType.String() == ResourceGroupResourceType.String() {
+			return nil, fmt.Errorf("invalid resource id: %s", id)
+		}
+
+		return newResourceIdentifier(parent, parts[0], ""), nil
+	}
+
+	if lowerFirstPart == providersKey && (len(parts) == 2 || strings.ToLower(parts[2]) == providersKey) {
+		//provider resource can only be on a tenant or a subscription parent
+		if parent.resourceType.String() != SubscriptionResourceType.String() && parent.resourceType.String() != TenantResourceType.String() {
+			return nil, fmt.Errorf("invalid resource id: %s", id)
+		}
+
+		return appendNext(newResourceIdentifierWithResourceType(parent, ProvidersResourceType, parts[1]), parts[2:], id)
+	}
+
+	if len(parts) > 3 && strings.ToLower(parts[0]) == providersKey {
+		return appendNext(newResourceIdentifierWithProvider(parent, parts[1], parts[2], parts[3]), parts[4:], id)
+	}
+
+	if len(parts) > 1 && strings.ToLower(parts[0]) != providersKey {
+		return appendNext(newResourceIdentifier(parent, parts[0], parts[1]), parts[2:], id)
+	}
+
+	return nil, fmt.Errorf("invalid resource id: %s", id)
+}
+
+func (id ResourceIdentifier) String() string {
+	if id.stringValue != nil {
+		return *id.stringValue
+	}
+
+	id.stringValue = to.StringPtr(id.toResourceString())
+	return *id.stringValue
+}
+
+func (id ResourceIdentifier) toResourceString() string {
+	if id.parent == nil {
+		return ""
+	}
+
+	builder := strings.Builder{}
+	builder.WriteString(id.parent.String())
+
+	if id.isChild {
+		builder.WriteString(fmt.Sprintf("/%s", id.resourceType.LastType()))
+		if len(id.name) > 0 {
+			builder.WriteString(fmt.Sprintf("/%s", id.name))
+		}
+	} else {
+		builder.WriteString(fmt.Sprintf("/providers/%s/%s/%s", id.resourceType.namespace, id.resourceType.t, id.name))
+	}
+
+	return builder.String()
+}
+
+func splitStringAndOmitEmpty(v, sep string) []string {
+	r := make([]string, 0)
+	for _, s := range strings.Split(v, sep) {
+		if len(s) == 0 {
+			continue
+		}
+		r = append(r, s)
+	}
+
+	return r
+}
+
+type ResourceType struct {
+	namespace string
+	t         string
+	types     []string
+
+	stringValue string
+}
+
+func (t ResourceType) Namespace() string {
+	return t.namespace
+}
+
+func (t ResourceType) Type() string {
+	return t.t
+}
+
+func (t ResourceType) LastType() string {
+	return t.types[len(t.types)-1]
+}
+
+func (t ResourceType) String() string {
+	return t.stringValue
+}
+
+func NewResourceType(providerNamespace, name string) ResourceType {
+	return ResourceType{
+		namespace:   providerNamespace,
+		t:           name,
+		types:       splitStringAndOmitEmpty(name, "/"),
+		stringValue: fmt.Sprintf("%s/%s", providerNamespace, name),
+	}
+}
+
+func NewResourceTypeFromParent(parent ResourceType, childType string) ResourceType {
```

Comment:
I wonder if this should be a method on `ResourceType`, did you try it to see how it works?

Reply:
To be a method will also work perfectly. Let me change this

Reply:
I have changed this one to a method `AppendChild`, please take a look